File: mcp/test/client.test.mjs
// 端到端测试：用假 Yearning 服务端验证 TS 客户端的 HTTP + WebSocket(msgpack) 往返。
// 运行：npm test（会先 build 出 dist/，再用 node:test 跑本文件）。

import test from "node:test";
import assert from "node:assert/strict";
import http from "node:http";
import { WebSocketServer } from "ws";
import { encode, decode } from "@msgpack/msgpack";
import { YearningClient } from "../dist/client.js";

function startFakeYearning() {
  let wsTokenSeen = null;
  let wsRefSeen = null;

  const server = http.createServer((req, res) => {
    const url = new URL(req.url, "http://x");
    let body = "";
    req.on("data", (c) => (body += c));
    req.on("end", () => {
      const json = (obj, status = 200) => {
        res.statusCode = status;
        res.setHeader("content-type", "application/json");
        res.end(JSON.stringify(obj));
      };
      if (url.pathname === "/login") {
        json({ code: 1200, payload: { token: "tok-123", real_name: "管理员", user: "admin" }, text: "" });
      } else if (url.pathname === "/api/v2/fetch/source") {
        assert.equal(url.searchParams.get("tp"), "query", "应通过 ?tp=query 传参");
        assert.equal(req.headers["authorization"], "Bearer tok-123");
        json({ code: 1200, payload: [{ source: "主库", idc: "bj", source_id: "s1" }], text: "" });
      } else if (url.pathname === "/api/v2/query/schema") {
        json({ code: 1200, payload: [{ title: "testdb", key: "testdb", meta: "Schema", isLeaf: false }], text: "" });
      } else if (url.pathname === "/api/v2/query/tables") {
        json({ code: 1200, payload: { table: [{ title: "account", key: "`testdb`.`account`", meta: "Table", isLeaf: true }] }, text: "" });
      } else if (url.pathname === "/api/v2/query/post") {
        res.statusCode = 200;
        res.end(); // 关闭审核时的空 body
      } else {
        res.statusCode = 404;
        res.end();
      }
    });
  });

  let wsOriginSeen = null;
  const wss = new WebSocketServer({ server, path: "/api/v2/query/results" });
  wss.on("connection", (ws, req) => {
    wsTokenSeen = req.headers["sec-websocket-protocol"] || null;
    wsOriginSeen = req.headers["origin"] || null;
    ws.on("message", (data) => {
      wsRefSeen = decode(data);
      const out = encode({
        query_time: 5,
        results: [{ field: [{ title: "n" }], data: [{ n: "1" }] }],
      });
      ws.send(out);
    });
  });

  return new Promise((resolve) => {
    server.listen(0, () => {
      const port = server.address().port;
      resolve({
        server,
        url: `http://127.0.0.1:${port}`,
        token: () => wsTokenSeen,
        origin: () => wsOriginSeen,
        ref: () => wsRefSeen,
      });
    });
  });
}

function makeJWT(exp) {
  const b64 = (o) => Buffer.from(JSON.stringify(o)).toString("base64url");
  return `${b64({ alg: "HS256" })}.${b64({ exp })}.sig`;
}

// 一个只关心鉴权的最小服务端：/login 计数并发 fresh token；受保护接口要求 Bearer fresh。
function startAuthServer() {
  let loginCount = 0;
  const server = http.createServer((req, res) => {
    let body = "";
    req.on("data", (c) => (body += c));
    req.on("end", () => {
      const url = new URL(req.url, "http://x");
      const json = (o, s = 200) => {
        res.statusCode = s;
        res.setHeader("content-type", "application/json");
        res.end(JSON.stringify(o));
      };
      if (url.pathname === "/login") {
        loginCount++;
        json({ code: 1200, payload: { token: "fresh", user: "admin" }, text: "" });
      } else if (url.pathname === "/api/v2/fetch/source") {
        if (req.headers["authorization"] !== "Bearer fresh") {
          res.statusCode = 401;
          res.end();
          return;
        }
        json({ code: 1200, payload: [{ source_id: "s1", source: "x", idc: "y" }], text: "" });
      } else {
        res.statusCode = 404;
        res.end();
      }
    });
  });
  return new Promise((resolve) =>
    server.listen(0, () => resolve({ server, url: `http://127.0.0.1:${server.address().port}`, count: () => loginCount }))
  );
}

test("401 自动重登重试", async () => {
  const a = await startAuthServer();
  try {
    const c = new YearningClient(a.url, "stale-token", { username: "admin", password: "pw", ldap: false });
    const s = await c.listSources();
    assert.equal(s.length, 1);
    assert.equal(s[0].source_id, "s1");
    assert.equal(a.count(), 1, "应触发一次自动登录");
  } finally {
    a.server.close();
  }
});

test("过期 token 主动重登（不等 401）", async () => {
  const a = await startAuthServer();
  try {
    const expired = makeJWT(Math.floor(Date.now() / 1000) - 3600);
    const c = new YearningClient(a.url, expired, { username: "admin", password: "pw", ldap: false });
    await c.listSources();
    assert.equal(a.count(), 1, "应在发请求前就主动登录一次");
  } finally {
    a.server.close();
  }
});

test("login + listSources（query 参数 + Bearer）", async () => {
  const fake = await startFakeYearning();
  try {
    const c = new YearningClient(fake.url);
    await c.login("admin", "pass");
    assert.equal(c.hasToken, true);
    const sources = await c.listSources();
    assert.equal(sources.length, 1);
    assert.equal(sources[0].source_id, "s1");
  } finally {
    fake.server.close();
  }
});

test("listSchemas / listTables", async () => {
  const fake = await startFakeYearning();
  try {
    const c = new YearningClient(fake.url, "tok-123");
    assert.deepEqual(await c.listSchemas("s1"), ["testdb"]);
    assert.deepEqual(await c.listTables("s1", "testdb"), ["account"]);
  } finally {
    fake.server.close();
  }
});

test("query：WebSocket msgpack 往返 + token 经 Sec-WebSocket-Protocol", async () => {
  const fake = await startFakeYearning();
  try {
    const c = new YearningClient(fake.url, "tok-123");
    const results = await c.query("s1", "testdb", "SELECT 1 AS n");
    assert.equal(fake.token(), "tok-123", "服务端应通过 Sec-WebSocket-Protocol 收到 token");
    assert.equal(fake.origin(), fake.url, "应携带 Origin（避免反代/WAF 403）");
    assert.deepEqual(fake.ref(), { type: 0, sql: "SELECT 1 AS n", schema: "testdb", source_id: "s1" });
    assert.equal(results.length, 1);
    assert.deepEqual(results[0].columns, ["n"]);
    assert.deepEqual(results[0].rows, [{ n: "1" }]);
  } finally {
    fake.server.close();
  }
});

File: cli/internal/client/client_test.go
package client

import (
	"encoding/base64"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/vmihailenco/msgpack/v5"
	"golang.org/x/net/websocket"
)

// makeJWT 构造一个仅 exp 有效的 JWT（签名无所谓，客户端只本地解析 exp）。
func makeJWT(exp int64) string {
	enc := func(v interface{}) string {
		b, _ := json.Marshal(v)
		return base64.RawURLEncoding.EncodeToString(b)
	}
	return enc(map[string]string{"alg": "HS256"}) + "." + enc(map[string]int64{"exp": exp}) + ".sig"
}

// 与后端一致的请求结构，用于在测试服务端解码客户端发来的查询。
type serverRef struct {
	Type     int    `msgpack:"type"`
	Sql      string `msgpack:"sql"`
	Schema   string `msgpack:"schema"`
	SourceId string `msgpack:"source_id"`
}

func TestLoginAndSources(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/login", func(w http.ResponseWriter, r *http.Request) {
		var body map[string]string
		_ = json.NewDecoder(r.Body).Decode(&body)
		if body["username"] != "admin" || body["password"] != "pass" {
			writeResp(w, Resp{Code: CodeLoginFail, Text: "用户名或密码错误"})
			return
		}
		writeResp(w, Resp{Code: CodeSuccess, Payload: mustJSON(LoginPayload{
			Token: "tok-123", RealName: "管理员", User: "admin", IsRecord: 1,
		})})
	})
	mux.HandleFunc("/ldap", func(w http.ResponseWriter, r *http.Request) {
		var body map[string]string
		_ = json.NewDecoder(r.Body).Decode(&body)
		if body["username"] != "ldapuser" {
			writeResp(w, Resp{Code: CodeLoginFail, Text: "LDAP 认证失败"})
			return
		}
		writeResp(w, Resp{Code: CodeSuccess, Payload: mustJSON(LoginPayload{
			Token: "ldap-tok", RealName: "LDAP 用户", User: "ldapuser",
		})})
	})
	mux.HandleFunc("/api/v2/fetch/source", func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("Authorization"); got != "Bearer tok-123" {
			t.Errorf("Authorization 头错误: %q", got)
		}
		writeResp(w, Resp{Code: CodeSuccess, Payload: mustJSON([]Source{
			{Source: "主库", IDC: "bj", SourceId: "mysql_01"},
		})})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	c := New(srv.URL, "")
	p, err := c.Login("admin", "pass")
	if err != nil {
		t.Fatalf("登录失败: %v", err)
	}
	if p.Token != "tok-123" || c.Token != "tok-123" {
		t.Fatalf("token 未正确保存: %+v", p)
	}

	sources, err := c.ListSources()
	if err != nil {
		t.Fatalf("获取数据源失败: %v", err)
	}
	if len(sources) != 1 || sources[0].SourceId != "mysql_01" {
		t.Fatalf("数据源解析错误: %+v", sources)
	}

	// 错误密码应返回明确错误。
	if _, err := New(srv.URL, "").Login("admin", "wrong"); err == nil {
		t.Fatal("错误密码应当登录失败")
	}

	// LDAP 登录走 /ldap 接口，契约一致。
	lp, err := New(srv.URL, "").LoginLDAP("ldapuser", "pw")
	if err != nil {
		t.Fatalf("LDAP 登录失败: %v", err)
	}
	if lp.Token != "ldap-tok" || lp.User != "ldapuser" {
		t.Fatalf("LDAP 登录解析错误: %+v", lp)
	}
}

func TestAutoReloginOn401(t *testing.T) {
	var loginCount int
	mux := http.NewServeMux()
	mux.HandleFunc("/login", func(w http.ResponseWriter, r *http.Request) {
		loginCount++
		writeResp(w, Resp{Code: CodeSuccess, Payload: mustJSON(LoginPayload{Token: "fresh", User: "admin"})})
	})
	mux.HandleFunc("/api/v2/fetch/source", func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Authorization") != "Bearer fresh" {
			w.WriteHeader(http.StatusUnauthorized)
			return
		}
		writeResp(w, Resp{Code: CodeSuccess, Payload: mustJSON([]Source{{SourceId: "s1"}})})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	var saved string
	c := NewWithAuth(srv.URL, "stale-token", &Credentials{Username: "admin", Password: "pw"}, func(tok string, exp int64) {
		saved = tok
	})
	sources, err := c.ListSources()
	if err != nil {
		t.Fatalf("应在 401 后自动重登成功: %v", err)
	}
	if len(sources) != 1 || sources[0].SourceId != "s1" {
		t.Fatalf("结果不符: %+v", sources)
	}
	if loginCount != 1 {
		t.Fatalf("应触发一次自动登录，实际 %d", loginCount)
	}
	if saved != "fresh" {
		t.Fatalf("onToken 应回传新 token，实际 %q", saved)
	}
}

func TestProactiveReloginOnExpiry(t *testing.T) {
	var loginCount int
	mux := http.NewServeMux()
	mux.HandleFunc("/login", func(w http.ResponseWriter, r *http.Request) {
		loginCount++
		writeResp(w, Resp{Code: CodeSuccess, Payload: mustJSON(LoginPayload{Token: "fresh", User: "admin"})})
	})
	mux.HandleFunc("/api/v2/fetch/source", func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Authorization") != "Bearer fresh" {
			w.WriteHeader(http.StatusUnauthorized)
			return
		}
		writeResp(w, Resp{Code: CodeSuccess, Payload: mustJSON([]Source{{SourceId: "s1"}})})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	// 传入一个已过期的 JWT：应在发请求“前”就主动重登（而非等 401）。
	expired := makeJWT(time.Now().Add(-time.Hour).Unix())
	c := NewWithAuth(srv.URL, expired, &Credentials{Username: "admin", Password: "pw"}, nil)
	if _, err := c.ListSources(); err != nil {
		t.Fatalf("过期应主动重登: %v", err)
	}
	if loginCount != 1 {
		t.Fatalf("应恰好主动登录一次，实际 %d", loginCount)
	}
}

func TestEnsureQueryOrderEmptyBody(t *testing.T) {
	// 关闭查询审核时，后端 ReferQueryOrder 成功后返回空 body。
	mux := http.NewServeMux()
	mux.HandleFunc("/api/v2/query/post", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK) // 不写任何 body
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	if err := New(srv.URL, "tok").EnsureQueryOrder("src", "reason"); err != nil {
		t.Fatalf("空响应体应视为成功，却报错: %v", err)
	}
}

func TestQueryWebSocket(t *testing.T) {
	const token = "ws-token-abc"
	handler := websocket.Handler(func(ws *websocket.Conn) {
		// 校验 token 通过 Sec-WebSocket-Protocol 传递。
		if got := ws.Request().Header.Get("Sec-WebSocket-Protocol"); got != token {
			t.Errorf("Sec-WebSocket-Protocol 头错误: %q", got)
		}
		var b []byte
		if err := websocket.Message.Receive(ws, &b); err != nil {
			t.Errorf("服务端接收失败: %v", err)
			return
		}
		var ref serverRef
		if err := msgpack.Unmarshal(b, &ref); err != nil {
			t.Errorf("服务端解码失败: %v", err)
			return
		}
		if ref.Sql != "SELECT 1 AS n" || ref.Schema != "testdb" {
			t.Errorf("查询内容错误: %+v", ref)
		}
		out, _ := msgpack.Marshal(QueryResults{
			QueryTime: 7,
			Results: []*Query{{
				Field: []map[string]interface{}{{"title": "n"}},
				Data:  []map[string]interface{}{{"n": "1"}},
			}},
		})
		_ = websocket.Message.Send(ws, out)
	})
	srv := httptest.NewServer(handlerWithSource(handler))
	defer srv.Close()

	c := New(srv.URL, token)
	res, err := c.Query("mysql_01", "testdb", "SELECT 1 AS n")
	if err != nil {
		t.Fatalf("查询失败: %v", err)
	}
	if res.Error != "" || res.Status {
		t.Fatalf("查询返回异常: %+v", res)
	}
	if len(res.Results) != 1 || len(res.Results[0].Data) != 1 {
		t.Fatalf("结果集解析错误: %+v", res.Results)
	}
	if got := res.Results[0].Data[0]["n"]; got != "1" {
		t.Fatalf("数据值错误: %v", got)
	}
	if cols := res.Results[0].Columns(); len(cols) != 1 || cols[0] != "n" {
		t.Fatalf("列解析错误: %v", cols)
	}
}

// handlerWithSource 把 websocket handler 挂到 /api/v2/query/results 路径。
func handlerWithSource(h http.Handler) http.Handler {
	mux := http.NewServeMux()
	mux.Handle("/api/v2/query/results", h)
	return mux
}

func writeResp(w http.ResponseWriter, r Resp) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(r)
}

func mustJSON(v interface{}) json.RawMessage {
	b, _ := json.Marshal(v)
	return b
}
